### PR TITLE
initial commit of dsky_clock

### PR DESCRIPTION
**Pull request by carlchan**
# DSKY Clock

This is a clockface inspired by the Apollo DSKY interface.

## Features
- battery level indicator (PROG indicator)
- month (VERB indicator)
- date (NOUN indicator)
- Current local time (DATA 1)
- UTC time (DATA 2)
- Step counter (DATA 3)

Indicator lights are:
- COMP ACTY (GPS/HRM active)
- MSG (Messages waiting)
- LOCK (Screen is locked)
- BT (Bluetooth is disconnected)
- BATT (Low battery or battery charging)
- ALARM (Alarm is set)
- STEP (Reached STEP goal))

----
This is my first watch face, may add features and customization later.

### CODE REVIEW

File: apps/dsky_clock/dsky_clock.app.js
Comment (thyttan) [inline apps/dsky_clock/dsky_clock.app.js:1]: Change name from `dsky_clock.app.js` to `app.js` to clear the sanitychecker.

Comment (carlchan) [reply apps/dsky_clock/dsky_clock.app.js:1]: Like this?

File: apps/dsky_clock/metadata.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{ "id": "dsky_clock",
+    "name": "DSKY Clock",
+    "shortName":"dsky_clock",
```
Comment (thyttan) [inline apps/dsky_clock/metadata.json:3]:
```suggestion
```

no need for a `shortName`, `name` is already short.

File: apps/dsky_clock/metadata.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    "screenshots": [{"url":"screenshot1.png"}],
+    "version": "0.01",
+    "description": "A clockface inspired by the Apollo DSKY interface.",
+    "tags": "clock",
+    "type": "clock",
+    "supports":["BANGLEJS2"],
+    "readme": "README.md",
+    "allow_emulator": true,
+    "storage": [
+      {"name":"app.js","url":"app.js"},
```
Comment (thyttan) [inline apps/dsky_clock/metadata.json:14]:
```suggestion
      {"name":"dsky_clock.app.js","url":"app.js"},
```

(`name` is the filename that will be set on the watch, and `url` is the filename in the BangleApps repo)

File: apps/dsky_clock/metadata.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+      {"name":"dsky_clock.img","url":"app-icon.js","evaluate":true}
```
Comment (thyttan) [inline apps/dsky_clock/metadata.json:15]: You'll have to add an `app-icon.js` file to the files in the repo dsky_clock folder - currently this links to non-existing file.

File: apps/dsky_clock/metadata.json [hunk 4]
```diff
@@ -0,0 +1,17 @@
+{ "id": "dsky_clock",
+    "name": "DSKY Clock",
+    "shortName":"dsky_clock",
+    "icon": "dsky_clock.img",
```
Comment (thyttan) [inline apps/dsky_clock/metadata.json:4]: I'd prefer if the icon name was changed to `app.png` to keep to the standard naming of the repo. I believe it has to be a png at least.

File: apps/dsky_clock/metadata.json (hunk 1 above)
Comment (thyttan) [reply apps/dsky_clock/metadata.json:3]: I added a new review comment with some more instructions for this :)

File: apps/dsky_clock/dsky_clock.app.js
Comment (thyttan) [reply apps/dsky_clock/dsky_clock.app.js:1]: I added a new review comment with some more instructions for this :)

File: apps/dsky_clock/metadata.json (hunk 4 above)
Comment (carlchan) [reply apps/dsky_clock/metadata.json:3]: I was following the icon section here:
https://github.com/espruino/EspruinoDocs/blob/master/tutorials/Bangle.js%20First%20App.md

when I was testing, if I just put the PNG then the icon doesn't seem to show up in my menu.  Happy to rename and/or upload the png version if needed tho?

Sorry, first time noob questions.

Comment (thyttan) [reply apps/dsky_clock/metadata.json:3]: Yes the js-coded icon is used for the watch (seems to be the one you have, I misinterpreted. So it should be renamed to `app-icon.js` in the repo) but the png is used for the app loader. 

More here: https://www.espruino.com/Bangle.js+App+Loader

Reading that now I found it a bit hard to follow - maybe some steps have been swapped around. But look and see if it helps! :)

File: apps/dsky_clock/app.js [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+      ]},
+    {type:"",width:5},
+  ],
+  lazy:true});
+layout.update();
+
+//support functions
+
+function getWeather() {
+  var weather = require("Storage").readJSON('weather.json', 1).weather;
```
Comment (thyttan) [inline apps/dsky_clock/app.js:63]: I don't have the weather app installed so I trigger this: 

```terminal
Uncaught Error: Can't read property 'weather' of undefined
 at line 70 col 63 in dsky_clock.app.js
...readJSON('weather.json', 1).weather;
                              ^
in function "getWeather" called from line 281 col 26 in dsky_clock.app.js
let weather=getWeather();
                       ^
in function "mode_weather" called from line 247 col 20 in dsky_clock.app.js
      mode_weather();
                   ^
in function "drawAlt" called from line 361 col 15 in dsky_clock.app.js
  drawAlt(mode);
              ^
in function called from system
> 
```

So some way of handling this situation is needed.

File: apps/dsky_clock/app.js [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+  drawAlt(mode);
+});
+
+g.clear();
+draw_bg();
+
+// Show launcher when middle button pressed
+Bangle.setUI({
+  mode : "clock",
+  remove : function() {
```
Comment (thyttan) [inline apps/dsky_clock/app.js:363]: You don't need a remove function if you're not implementing fast loading so I'd say just remove it: https://www.espruino.com/Bangle.js+Fast+Load

I would suggest to skip that for now. You could add it later if you wanted to. :+1:

File: apps/dsky_clock/app.js [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+  if (mode < 0 ) { mode=5; }
+  mode=(mode % 6);
+  drawAlt(mode);
+});
+
+g.clear();
+draw_bg();
+
+// Show launcher when middle button pressed
+Bangle.setUI({
```
Comment (thyttan) [inline apps/dsky_clock/app.js:361]: I wonder if it would be better to go without Bangle.setUI and just work with the layout library. But I'm not sure myself.

It doesn't make the button open the launcher for me. Maybe that is a conflict with the layout library?

File: apps/dsky_clock/app.js [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+  mode : "clock",
+  remove : function() {
+    if (drawTimeout) clearTimeout(drawTimeout);
+    if (AltDrawTimer) clearTimeout(AltDrawTimer);
+    drawTimeout = undefined;
+    AltDrawTimer = undefined;
+    require("widget_utils").show(); // re-show widgets
+  }});
+Bangle.loadWidgets();
+require("widget_utils").swipeOn(); // hide widgets, make them visible with a swipe
```
Comment (thyttan) [inline apps/dsky_clock/app.js:371]: For some reason the widgets are not off screen for me. Don't know why since I'd expect this `swipeOn()` call to hide them.

File: apps/dsky_clock/app.js (hunk 7 above)
Comment (carlchan) [reply apps/dsky_clock/app.js:366]: Yes, looks like it breaks when I set layout.setUI() from the drawMain function.  Commented it out for now.

File: apps/dsky_clock/app.js (hunk 6 above)
Comment (carlchan) [reply apps/dsky_clock/app.js:368]: I had no idea I was doing that :p switching to plain non-fast-load for now since I doubt I'm properly removing various events etc

File: apps/dsky_clock/app.js (hunk 8 above)
Comment (carlchan) [reply apps/dsky_clock/app.js]: think this was related to the layout setUI issues too, seems to work now.

**Comment by thyttan** [comment]
Some of the ui-elements seem to be drawn a bit further down compared to the icon on my watch:
image

**Comment by thyttan** [comment]
Refactored a bit and PR'd to your fork. Use it if you want! https://github.com/carlchan/BangleApps/pull/1

**Comment by carlchan** [comment]
> Refactored a bit and PR'd to your fork. Use it if you want! https://github.com/carlchan/BangleApps/pull/1

Fantastic! Thank you!

**Comment by thyttan** [comment]
If you want to we could merge now. Just say! :)

I'd suggest for a next version to look at another background color for the indicator field on the left that doesn't need to be dithered - since dithering massively hinders legibility of the text on the watch!

**Comment by carlchan** [comment]
If you think it's ready, then yes please!

**Comment by thyttan** [comment]
Let's go ahead then - thanks for contributing your first app! 🎉
